dimod/binary/binary_quadratic_model.py

```diff
@@ -390,30 +392,161 @@ def __deepcopy__(self, memo):
         memo[id(self)] = new
         return new
 
-    def __eq__(self, other):
-        # todo: performance
-        try:
-            return (self.vartype == other.vartype
-                    and self.shape == other.shape  # redundant, fast to check
-                    and self.offset == other.offset
-                    and self.linear == other.linear
-                    and self.adj == other.adj)
-        except AttributeError:
-            return False
-
     def __len__(self):
         return self.num_variables
 
-    def __ne__(self, other):
-        return not self == other
-
     def __repr__(self):
         return "{!s}({!s}, {!s}, {!r}, {!r})".format(type(self).__name__,
                                                      self.linear,
                                                      self.quadratic,
                                                      self.offset,
                                                      self.vartype.name)
 
+    def __add__(self, other: Union['BinaryQuadraticModel', Bias]):
+        # in python 3.8+ we could do this is functools.singledispatchmethod
+        if isinstance(other, BinaryQuadraticModel):
+            if other.num_variables and other.vartype != self.vartype:
+                # future: return QuadraticModel
+                raise TypeError("cannot add BQMs with different vartypes")
+            new = self.copy()
+            new.update(other)
+            return new
+        if isinstance(other, Number):
```

Comment:
Doesn't `DictBQM` support `object` for bias `dtype`?

Currently `NotImplemented` is raised, even if that object implements `+`.

In other places as well.

Reply:
Hmm, you're right. And I just checked that `isinstance(sympy.Symbol('a'), Number) == False`... I'll take a look.

Reply:
Ok, I played with a couple different variations of trying to handle this case and this way lies madness. I think (for now) only allowing symbolic manipulation for numeric types strikes a good balance between nice sugar and not having to deal with a million edge cases.

Users can always make use of the `bqm.offset += val` and `bqm.scale(val)` directly, which will handle other types reasonably well (at least when combined with #624 )

Reply:
Yeah, among other things, a bias can be just another BQM :laughing: